[Merged by Bors] - chore(ring_theory/dedekind_domain): speed up `dedekind_domain.lean`

Vierkantor opened the pull request:
@eric-wieser [noticed that `dedekind_domain.lean`](https://leanprover.zulipchat.com/#narrow/stream/113488-general/topic/Timeouts.20in.20ring_theory.2Fdedekind_domain.2Elean.3A664.3A9) was compiling slowly and on the verge of a timeout. @kbuzzard, @sgouezel and I reworked some definitions to make everything elaborate much faster: `is_dedekind_domain_inv_iff`, `mul_inv_cancel_of_le_one` and `ideal.unique_factorization_monoid` went from over 10 seconds on my machine to less than 3 seconds. No other declaration in that file now takes over 2 seconds on my machine.

Apart from the three declarations getting new proofs, I also made the following changes:
 * The operations on `localization` (`has_add`, `has_mul`, `has_one`, `has_zero`, `has_neg`, `npow` and `localization.inv`) are now `@[irreducible]`
 * `fraction_ring.field` copies its field from `localization.comm_ring` for faster unification (less relevant after the previous change)
 * Added `fractional_ideal.map_mem_map` and `fractional_ideal.map_injective` to simplify the proof of `is_dedekind_domain_inv_iff`.
 * Split the proof of `matrix.exists_mul_vec_eq_zero_iff` into two parts to speed it up

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

Vierkantor committed:
Make the `localization` operations irreducible

Vierkantor committed:
Speed up the proof for `is_dedekind_domain_iff`

Vierkantor committed:
Remove the `by_contradiction` from `mul_inv_cancel_of_le_one`

Vierkantor committed:
Speed up `ideal.unique_factorization_monoid`

Co-Authored-By: sgouezel <sebastien.gouezel@univ-rennes1.fr>
Co-Authored-By: Kevin Buzzard <k.buzzard@imperial.ac.uk>

jcommelin reviewed (approved):
Thanks :tada:

bors d+

eric-wieser commented on src/ring_theory/localization.lean:
```diff
@@ -662,7 +662,9 @@ open is_localization
-instance : has_add (localization M) :=
+section
+
+@[irreducible] instance : has_add (localization M) :=
```
I don't think this is a good idea. I think the strategy is normally to make a `protected def add`, and make _that_ irreducible (which is what we do for polynomial)

Vierkantor replied:
Sure, I'll do that. But why is this a bad idea specifically?

eric-wieser replied:
It's a bad idea because I think typeclass search is unable to unify `localization.has_add.add = localization.add_monoid.to_has_add.add` as doing so would require it to reduce `localization.has_add` and look at the structure fields.

Vierkantor replied:
I tried moving them into irreducible defs, but now I get an error in `localization.lean`. Changing `monoid_of` so it didn't unfold the irreducible defs didn't help either.

Vierkantor replied:
Finally, making even more things irreducible works!

Vierkantor committed:
Make new protected defs

Vierkantor committed:
Make `localization` an irreducible def

This should ensure the API for `localization` is contained within `mk`/`mk_eq_mk_iff`/`rec`, no more `quotient` defs escaping

Vierkantor committed:
Make `localization.npow` irreducible too

This halves the elaboration time of `matrix.mul_vec_eq_zero_iff`

Vierkantor committed:
Fix

Vierkantor committed:
Add docstrings

Vierkantor committed:
Lint fix: add `simp`-normal forms for `lift_on_mk` and `lift_on₂_mk`

Vierkantor committed:
Add more docstrings

Vierkantor commented on src/group_theory/monoid_localization.lean:
```diff
@@ -175,8 +175,53 @@ namespace localization
+This is a separate `irreducible` def to ensure the elaborator doesn't waste its time
+trying to unify some huge recursive definition with itself, but unfolded one step less.
+-/
+@[to_additive "Multiplication with a natural in an `add_localization` is defined as `n • ⟨a, b⟩ = ⟨n • a, n • b⟩`.
```
```suggestion
@[to_additive
"Multiplication with a natural in an `add_localization` is defined as `n • ⟨a, b⟩ = ⟨n • a, n • b⟩`.
```

Vierkantor commented on src/group_theory/monoid_localization.lean:
```diff
@@ -175,8 +175,53 @@ namespace localization
+@[to_additive "Multiplication with a natural in an `add_localization` is defined as `n • ⟨a, b⟩ = ⟨n • a, n • b⟩`.
+
+This is a separate `irreducible` def to ensure the elaborator doesn't waste its time
+trying to unify some huge recursive definition with itself, but unfolded one step less.", irreducible]
```
```suggestion
trying to unify some huge recursive definition with itself, but unfolded one step less.",
irreducible]
```

Vierkantor committed:
Fix overlong lines
